File: packages/kotlin/src/test/kotlin/com/ecollect/sdk/PaymentLinksModuleTest.kt
package com.ecollect.sdk

import com.ecollect.sdk.exceptions.ValidationException
import com.ecollect.sdk.modules.PaymentLinkMethod
import com.ecollect.sdk.modules.PaymentLinksModule
import com.ecollect.sdk.modules.PaymentsModule
import com.ecollect.sdk.modules.SessionModule
import com.ecollect.sdk.types.CreateTransactionResponse
import com.ecollect.sdk.types.PaymentIntent
import com.ecollect.sdk.utils.HttpClient
import io.mockk.*
import kotlinx.coroutines.test.runTest
import org.junit.Assert.*
import org.junit.Test

class PaymentLinksModuleTest {

    private val baseIntent = PaymentIntent(
        amount = 100.0,
        currency = "COP",
        referenceArray = listOf("CC", "12345678", "ORDER-001", "Juan Perez", "user@example.com", "3001234567"),
        usermail = "user@example.com",
        cardHolderName = "Juan Perez",
        cardHolderIdType = "CC",
        cardHolderId = "12345678"
    )

    private fun makeModule(): Triple<PaymentLinksModule, PaymentsModule, SessionModule> {
        val config = Config(etyCode = 123, sessionToken = "test-session", environment = Environment.TEST)
        val httpClient = mockk<HttpClient>()
        val session = mockk<SessionModule>()
        coEvery { session.getSessionToken() } returns "test-session"
        val payments = mockk<PaymentsModule>()
        val module = PaymentLinksModule(config, payments, session)
        return Triple(module, payments, session)
    }

    @Test
    fun `generatePaymentLink returns valid URL via email method`() = runTest {
        val (module, payments) = makeModule()
        coEvery { payments.process(any()) } returns CreateTransactionResponse(
            returnCode = "SUCCESS",
            ticketId = 42L,
            eCollectUrl = "https://pay.ecollect.com/link/abc123",
            lifetimeSecs = 3600
        )

        val result = module.generatePaymentLink(baseIntent, PaymentLinkMethod.EMAIL)

        assertEquals(42L, result.ticketId)
        assertEquals("https://pay.ecollect.com/link/abc123", result.eCollectUrl)
        assertEquals(3600, result.lifetimeSecs)
    }

    @Test
    fun `generatePaymentLink with QR method passes lifetime in extra info`() = runTest {
        val (module, payments) = makeModule()
        coEvery { payments.process(any()) } returns CreateTransactionResponse(
            returnCode = "SUCCESS",
            ticketId = 99L,
            eCollectUrl = "https://pay.ecollect.com/link/qr999",
            lifetimeSecs = 7200
        )

        val result = module.generatePaymentLink(baseIntent, PaymentLinkMethod.QR, lifetimeSecs = 7200)

        assertEquals(99L, result.ticketId)
        assertEquals(7200, result.lifetimeSecs)
    }

    @Test
    fun `generatePaymentLink SMS throws if mobileCountryCode is null`() = runTest {
        val (module) = makeModule()
        val intentNoMobile = baseIntent.copy(mobileCountryCode = null, mobileNumber = null)

        try {
            module.generatePaymentLink(intentNoMobile, PaymentLinkMethod.SMS)
            fail("Should throw IllegalArgumentException")
        } catch (e: IllegalArgumentException) {
            assertTrue(e.message?.contains("mobileCountryCode") == true)
        }
    }

    @Test
    fun `generatePaymentLink EMAIL throws if usermail is null`() = runTest {
        val (module) = makeModule()
        val intentNoEmail = baseIntent.copy(usermail = null)

        try {
            module.generatePaymentLink(intentNoEmail, PaymentLinkMethod.EMAIL)
            fail("Should throw IllegalArgumentException")
        } catch (e: IllegalArgumentException) {
            assertTrue(e.message?.contains("usermail") == true)
        }
    }

    @Test
    fun `generatePaymentLink sets paymentSystem to 10`() = runTest {
        val (module, payments) = makeModule()
        val capturedSlot = slot<PaymentIntent>()
        coEvery { payments.process(capture(capturedSlot)) } returns CreateTransactionResponse(
            returnCode = "SUCCESS",
            ticketId = 55L,
            eCollectUrl = "https://pay.ecollect.com/link/55",
            lifetimeSecs = 3600
        )

        module.generatePaymentLink(baseIntent, PaymentLinkMethod.EMAIL)

        assertEquals("10", capturedSlot.captured.paymentSystem)
    }
}

File: packages/php/tests/PaymentLinksModuleTest.php
<?php

declare(strict_types=1);

namespace Ecollect\Tests;

use Ecollect\Config;
use Ecollect\Exceptions\ValidationException;
use Ecollect\Modules\PaymentLinksModule;
use Ecollect\Modules\PaymentsModule;
use Ecollect\Modules\SessionModule;
use Ecollect\Types\Customer;
use Ecollect\Types\PaymentIntent;
use Ecollect\Utils\HttpClient;
use PHPUnit\Framework\MockObject\MockObject;
use PHPUnit\Framework\TestCase;

class PaymentLinksModuleTest extends TestCase
{
    private function makeConfig(): Config
    {
        return new Config([
            'api_key'     => 'test-key',
            'ety_code'    => 123,
            'environment' => 'test',
            'srv_code'    => 1,
        ]);
    }

    private function baseIntent(): PaymentIntent
    {
        return new PaymentIntent([
            'amount'   => 100.0,
            'currency' => 'COP',
            'customer' => new Customer([
                'email'           => 'user@example.com',
                'document_number' => '12345678',
                'document_type'   => 'CC',
                'full_name'       => 'Juan Perez',
            ]),
        ]);
    }

    /** @return PaymentsModule&MockObject */
    private function mockPaymentsWithResult(array $result): PaymentsModule
    {
        $http    = $this->createMock(HttpClient::class);
        $session = $this->createMock(SessionModule::class);
        $session->method('getActive')->willReturn('session-token');

        $payments = $this->getMockBuilder(PaymentsModule::class)
            ->setConstructorArgs([$this->makeConfig(), $http, $session])
            ->onlyMethods(['processWithExtraInfo'])
            ->getMock();

        $payments->method('processWithExtraInfo')->willReturn($result);

        return $payments;
    }

    public function testGeneratePaymentLinkReturnsValidUrl(): void
    {
        $payments = $this->mockPaymentsWithResult([
            'eCollectUrl'  => 'https://pay.ecollect.com/link/abc123',
            'ticketId'     => 42,
            'lifetimeSecs' => 3600,
        ]);

        $module = new PaymentLinksModule($payments);
        $result = $module->generatePaymentLink($this->baseIntent(), 'email');

        $this->assertSame('https://pay.ecollect.com/link/abc123', $result['eCollectUrl']);
        $this->assertSame(42, $result['ticketId']);
        $this->assertSame(3600, $result['lifetimeSecs']);
        $this->assertGreaterThan(time(), $result['expiresAt']);
    }

    public function testGeneratePaymentLinkWithOptionalFields(): void
    {
        $payments = $this->mockPaymentsWithResult([
            'eCollectUrl'  => 'https://pay.ecollect.com/link/qr999',
            'ticketId'     => 99,
            'lifetimeSecs' => 7200,
        ]);

        $intent                = $this->baseIntent();
        $intent->qrLifetimeSecs = 7200;

        $module = new PaymentLinksModule($payments);
        $result = $module->generatePaymentLink($intent, 'qr');

        $this->assertSame(7200, $result['lifetimeSecs']);
        $this->assertSame(99, $result['ticketId']);
    }

    public function testGeneratePaymentLinkSmsThrowsIfMobileDataMissing(): void
    {
        $payments = $this->mockPaymentsWithResult([]);
        $module   = new PaymentLinksModule($payments);

        $intent = $this->baseIntent(); // no mobileCountryCode / mobileNumber

        $this->expectException(ValidationException::class);
        $module->generatePaymentLink($intent, 'sms');
    }

    public function testGeneratePaymentLinkExpiredSessionStillReturnsResult(): void
    {
        // PaymentsModule handles session refresh internally; PaymentLinksModule
        // just delegates to processWithExtraInfo and maps the response.
        $payments = $this->mockPaymentsWithResult([
            'eCollectUrl'  => 'https://pay.ecollect.com/link/refresh-ok',
            'ticketId'     => 55,
            'lifetimeSecs' => 3600,
        ]);

        $module = new PaymentLinksModule($payments);
        $result = $module->generatePaymentLink($this->baseIntent(), 'email');

        $this->assertSame('https://pay.ecollect.com/link/refresh-ok', $result['eCollectUrl']);
    }
}
